src/realm/sync/network/default_socket.cpp

```diff
@@ -371,10 +378,172 @@ void DefaultWebSocketImpl::initiate_websocket_handshake()
+
+    auto will_destroy_thread = util::make_scope_exit([&]() noexcept {
+        m_logger_ptr->trace("Default event loop: thread exiting");
+        if (g_binding_callback_thread_observer)
+            g_binding_callback_thread_observer->will_destroy_thread();
+
+        {
+            std::lock_guard<std::mutex> lock(m_mutex);
+            REALM_ASSERT(m_state == State::Stopping);
+
+            // Reset the service so it can be run again later
+            m_service.reset();
+            do_state_update(State::Stopped);
+        }
+    });
+    if (g_binding_callback_thread_observer)
+        g_binding_callback_thread_observer->did_create_thread();
+
+    lock.lock();
+    if (m_state != State::Started) {
+        // Stop has already been requested - exit early
+        return;
+    }
+    do_state_update(State::Running);
+    lock.unlock();
+
+    m_logger_ptr->trace("Default event loop: service run");
+    try {
+        m_service.run(); // Throws
+    }
+    catch (const std::exception& e) {
+        m_logger_ptr->error("Default event loop exception: ", e.what());
+        if (g_binding_callback_thread_observer)
+            g_binding_callback_thread_observer->handle_error(e);
+        else
+            throw;
+    }
+}
+
+void DefaultSocketProvider::stop(bool wait_for_stop)
+{
+    std::unique_lock<std::mutex> lock(m_mutex);
+
+    // Do nothing if the thread is not started or running or stop has already been called
+    if (m_state == State::Starting || m_state == State::Started || m_state == State::Running) {
+        m_logger_ptr->trace("Default event loop: stop()");
+        do_state_update(State::Stopping);
+        // Updating state to Stopping will free a start() if it is waiting for the thread to
+        // start and may cause the thread to exit early before calling service.run()
+        m_service.stop(); // Unblocks m_service.run()
+    }
+
+    // Wait until the thread is stopped (exited) if requested
+    if (wait_for_stop && m_state == State::Stopping) {
+        lock.unlock();
+        m_logger_ptr->trace("Default event loop: wait for stop");
+        state_wait_for(State::Stopped);
+    }
+}
+
+//                                   +-------------------------------------------+
+//                                  \/                                           |
+// State Machine: NotStarted -> Starting -> Started -> Running -> Stopping -> Stopped
+//                                  |          |                     ^
+//                                  +----------+---------------------+
+
+void DefaultSocketProvider::do_state_update(State new_state)
+{
+    // m_state_mutex should already be locked...
+    m_state = new_state;
+    m_state_cv.notify_all(); // Let any waiters check the state
+}
+
+bool DefaultSocketProvider::state_wait_for(State expected_state)
+{
+    std::unique_lock<std::mutex> lock(m_mutex);
+    bool match = false;
+    // Check for condition already met
+    if (m_state == expected_state)
```

Comment:
shouldn't this be `if (m_state >= expected_state)`? (same below)

Reply:
The function returns true or false depending on whether the state matches exactly. Since the code isn't using the return value, I can remove it.